src/profile-logic/marker-data.js

```diff
@@ -522,78 +552,284 @@ export function deriveMarkersFromRawMarkerTable(
+          // non-overlapping parts of the same load. The start marker has a
+          // duration from channel-creation until Start (i.e. AsyncOpen()). The
+          // end marker has a duration from AsyncOpen time until OnStopRequest.
+          // In the merged marker, we want to represent the entire duration, from
+          // channel-creation until OnStopRequest.
+          //
+          // |--- start marker ---|--- stop marker with timings ---|
+          //
+          // Usually the start marker is very small. It's emitted mostly to know
+          // about the start of the request. But most of the interesting bits are
+          // in the stop marker.
+
+          const ensureMessage =
+            'Network markers are assumed to have a start and end time.';
+          if (data.status === 'STATUS_START') {
+            openNetworkMarkers.set(data.id, rawMarkerIndex);
+          } else {
+            // End status can be any status other than 'STATUS_START'. They are
+            // either 'STATUS_STOP' or 'STATUS_REDIRECT'.
+            const endData = data;
 
-  for (let i = 0; i < rawMarkers.length; i++) {
-    const name = rawMarkers.name[i];
-    // TODO - A follow-up commit will correctly handle startTime and endTime.
-    const time = ensureExists(rawMarkers.startTime[i], ensureExistsMessage);
-    const data = rawMarkers.data[i];
-    const category = rawMarkers.category[i];
-
-    if (!data) {
-      // Add a marker with a zero duration
-      matchedMarkers.push({
-        start: time,
-        dur: 0,
-        name: stringTable.getString(name),
-        title: null,
-        category,
-        data: null,
-      });
-      continue;
+            const startIndex = openNetworkMarkers.get(data.id);
+            if (startIndex !== undefined) {
+              // A start marker matches this end marker.
+              openNetworkMarkers.delete(data.id);
+
+              // We know this startIndex points to a Network marker.
+              const startData: NetworkPayload = (rawMarkers.data[
+                startIndex
+              ]: any);
+
+              const startStartTime = ensureExists(
+                rawMarkers.startTime[startIndex],
+                ensureMessage
+              );
+              const endStartTime = ensureExists(maybeStartTime, ensureMessage);
+              const endEndTime = ensureExists(maybeEndTime, ensureMessage);
+
+              markers.push({
+                start: startStartTime,
+                dur: endEndTime - startStartTime,
+                name: stringTable.getString(name),
+                title: null,
+                category,
+                data: {
+                  ...endData,
+                  startTime: startStartTime,
+                  fetchStart: endStartTime,
+                  cause: startData.cause || endData.cause,
+                },
+              });
+              markerIndexToRawMarkerIndexes.push([startIndex, rawMarkerIndex]);
```

Comment:
We are always using `markers.push` and `markerIndexToRawMarkerIndexes.push` at the same time. Do you think creating a local arrow function that pushes to these arrays would be helpful? It can make sure that we are adding both in the future, like when we add a custom marker that needs a custom handling.

Reply:
That would definitely help with copy/pasta programming errors.